### PR TITLE
Capture worker logs

### PR DESCRIPTION
**Pull request by drewcrawford**
# Problem

Calls to `console.log`debug, etc., from workers are not forwarded to console.  They are silently dropped.

This is mainly a problem when you are doing `log("GOT HERE")` style debugging, to determine whether or not you actually got there, and the place you got there is a worker, and you incorrectly conclude you didn't get there, because the log isn't forwarded to console.

# Solution

Monkeypatch Worker.new() so that, each time a worker is created, its console methods are also monkeypatched in a similar way to the main page, where the logs are forwarded to console.

# Alternatives

There's an obvious risk in monkeypatching more APIs.  There are a lot of edge cases in worker types: dedicated/service/shared, blob/url, etc.  Realistically I would need help to track all the edge cases down, hence the draft.

I sent an alternative, opt-in design to wasm-bindgen-test-runner at https://github.com/wasm-bindgen/wasm-bindgen/pull/4856.  Opt-in is safer but less satisfactory.

I wanted to prototype the riskier and better approach, and it occurred to me there might be a different appetite for risk in this runner.

### CODE REVIEW

**Comment by drewcrawford** [comment]
I have now verified this against my [log_party](https://github.com/drewcrawford/log_party) test suite.  After applying this PR, this is the first runner to hit 100%.

That seems like an appropriate time to request a review.

**Comment by jakobhellermann** [comment]
Thanks for the PR! This looks like a nice addition and is definitely in scope for the project.

I'll give it a review sometime over the next couple of days when I get to it.

**Comment by drewcrawford** [comment]
Thanks!  And take your time.

Beyond this instant PR, the broader ecosystem question of whether to do this pattern or the https://github.com/wasm-bindgen/wasm-bindgen/pull/4856 pattern seems to most of us like a tough call with complicated tradeoffs. 

Since "those who wrote their own wasm runner" is a very short list, I think your expertise would be very valuable in terms of identifying the right solution here.

File: static/index.html [hunk 1]
```diff
@@ -167,6 +167,91 @@
...
+                    event.stopImmediatePropagation();
+                }
+            });
+            worker.onerror = function(e) {
+                console.error(e.message || 'Worker error');
+            };
+        }
+
+        Worker = function(scriptURL, options) {
+            const resolvedURL = new URL(scriptURL, location.href).href;
```
Comment (jakobhellermann) [inline static/index.html:235]: Apparently `URL` can also take a TrustedScriptUrl instead of a string https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker. No idea if that would break this code.

File: static/index.html [hunk 2]
```diff
@@ -167,6 +167,91 @@
...
+            const isModule = options?.type === "module";
+
+            // For blob URLs, fetch content synchronously and prepend our shim
+            if (resolvedURL.startsWith("blob:")) {
+                const xhr = new XMLHttpRequest();
+                xhr.open("GET", resolvedURL, false);
+                xhr.send();
```
Comment (jakobhellermann) [inline static/index.html]: out of curiosity, it there a reason to use `XMLHttpRequest` instead of `fetch`? It's been a while since I have seen it 😄

File: static/index.html [hunk 3]
```diff
@@ -167,6 +167,91 @@
...
+                const __stdlog = console[__logMethod].bind(console);
+                console[__logMethod] = function (...args) {
+                    __stdlog(...args);
+                    postMessage({ __wasm_server_runner_log: true, level: __logMethod, args: args.map(a => {
+                        try {
+                            return (typeof a === "object" && a !== null) ? JSON.parse(JSON.stringify(a)) : a;
+                        } catch {
+                            return String(a);
+                        }
+                    })});
```
Comment (jakobhellermann) [inline static/index.html:184]: Why is this necessary, can't we just `postMessage({ args })`? I thought those would get passed using `structuredClone` which I would expect to be just superior to `JSON.stringify`.

**Review by jakobhellermann — APPROVED** [review]
I forgot to review this, but I just stumbled upon it again. I still think this is a nice addition, and looking at https://github.com/wasm-bindgen/wasm-bindgen/pull/4856 I would intuitively say that `wasm-server-runner` is a good place for this functionality.

> There's an obvious risk in monkeypatching more APIs. There are a lot of edge cases in worker types: dedicated/service/shared, blob/url, etc. Realistically I would need help to track all the edge cases down, hence the draft.

Nobody is using `wasm-server-runner` in production (I hope) so I think that it's totally fine to have something a bit hacky in here in order to improve the dev experience. I'm also down to merge this as is, and if any edge cases or issues pop up later, fix those as they come.

---

The implementation looks good to me, I did leave a couple of comments.

One thing I'd like to see is an example in `example-project/examples/worker.rs` that makes it easy to test this. 

I'd merge this either way but only with an example will I make sure that I don't accidentally break this when making an update 😄

**Comment by drewcrawford** [comment]
I empathize with the breakage risk completely; added new example to diff outputs

File: static/index.html (hunk 1 above)
Comment (drewcrawford) [reply static/index.html:235]: After looking into this, I made a minor change [here](https://github.com/jakobhellermann/wasm-server-runner/pull/58/changes#diff-3e8602c86d6cf04a5c74b954f9f957b652617102cf8f4b647fe46989f480861dR224) to address what seems like a very theoretical compatibility problem that might affect somebody somewhere, but I think it pretty much works as is:

1.  TrustedScriptUrl gets coerced down to its url/string form and should "just work" like any other arg.  Added example to test this argument type.
2. The reason one has a TrustedScriptUrl to pass in, is that one is expecting the CSP header `Content-Security-Policy: require-trusted-types-for 'script'` to show up (e.g., in production).  Since wasm-server-runner  does not set it, there's nothing to 'enforce' in that configuration.
3. Ultimately, injecting logging code is incompatible with the trusted script model, but that just comes down to a difference between development and production HTTP servers.

File: static/index.html (hunk 2 above)
Comment (drewcrawford) [reply static/index.html]: The original reason was to fetch synchronously.  However on reflection, I think we don't need to special-case this after all, so we can do away with the synchronous fetch as well.

File: static/index.html (hunk 3 above)
Comment (drewcrawford) [reply static/index.html:184]: It is superior when restricted to the subset of structured cloneables, so it makes sense to guess this as a possible optimization to some subset of possible inputs.  But it doesn't solve the general problem.

`console.log` should work not just for cloneables, but for the broader universe of non-cloneables, so we need a path that actually logs the non-cloneables, and we should consider focusing our efforts on one path there. The usual example of a non-cloneable type is a JS function, which is loggable but not cloneable. But the broader web apis includes a lot more (WebGPU types, AudioContext, RTC, sockets...) as well as containers (Array,Promise,Error,etc.) that may have some noncloneable in some slot.  

The real usecase, is that `console.error` and such are ultimately going to be called by some panic/exception handler, to log info about why that panicked or excepted.  Mainly, the last thing to do inside the logging inside a panic handler is to begin a brand new exception that we can't log the other one, because one of its slots was not cloneable.   What was the other exception anyway, such as to make it cloneable?  So the underlying idea behind my control flow here is "log what you can with best effort, and fall back to log something less detailed as necessary"

> can't we just `postMessage({ args })`

The main way to do this would be to do it as a try-catch block *on top of these changes*.  Reason I'm not proposing it out of the box is I haven't encountered it as a performance bottleneck yet.  In addition to the fact that this work at all, the other runner is the squeaky wheel for my performance benchmarks so far.
